File: art/audio/2026-09-26-release/live.mjs
#!/usr/bin/env node
/**
 * live.mjs — record the real graph running, so the cut is proved where the owner hears it.
 *
 *   node art/audio/2026-09-26-release/live.mjs --dist dist --out /tmp/live/after [--seconds 16]
 *
 * Every number in this folder comes from `renderOffline`: the same graph rebuilt in an
 * `OfflineAudioContext` and driven by a scripted pass against a perfect clock. That is exact, and
 * it is a TWIN. `SFX_TRIM` is set inside `createBuses`, which both paths share, so there is no
 * plausible way for one to have it and the other not — but "no plausible way" is an argument, and
 * this lane's rule is that an argument is not a measurement.
 *
 * So: the real build, in play mode, running north to south over the plaza flagstones with the
 * master bus recorded through a `MediaStreamDestination`. Run either side of the change and
 * compare the footstep peaks (`live.py`). The pulse itself is not measured here — sixteen seconds
 * is far too short a window to resolve a 1.2 Hz beat — only the level, which is what was changed.
 */
import fs from 'node:fs';
import path from 'node:path';
import { serveStatic, launchBrowser, READY_TIMEOUT_MS } from '../../../gauntlet/scripts/lib/browser.mjs';

const args = Object.fromEntries(
  process.argv
    .slice(2)
    .map((a, i, all) => (a.startsWith('--') ? [a.slice(2), all[i + 1]?.startsWith('--') || all[i + 1] === undefined ? true : all[i + 1]] : []))
    .filter(Boolean),
);
const dist = path.resolve(args.dist || 'dist');
const out = path.resolve(args.out || '/tmp/live/after');
const seconds = Number(args.seconds ?? 16);
const DT = 1 / 60;
const log = (...m) => console.error('[live]', ...m);
fs.mkdirSync(out, { recursive: true });

const server = await serveStatic(dist);
const browser = await launchBrowser({ width: 640, height: 360 });
try {
  const page = await browser.newPage();
  await page.evaluateOnNewDocument(() => Object.defineProperty(navigator, 'webdriver', { get: () => false }));
  page.on('pageerror', (e) => log('[pageerror]', e.message));
  await page.goto(`${server.url}/?test=1&dev=0&hud=0&warmup=0&quality=high`, { waitUntil: 'load', timeout: READY_TIMEOUT_MS });
  await page.waitForFunction(() => !!window.__ZR__ && !!window.__ZR_PLAY__, { timeout: READY_TIMEOUT_MS, polling: 500 });
  await page.evaluate(() => window.__ZR__.ready());
  log('world ready');
  // two taps on M: the shell starts muted in a headless page, and the second one unmutes
  for (let i = 0; i < 2; i++) {
    await page.keyboard.down('KeyM');
    await page.keyboard.up('KeyM');
  }
  await new Promise((r) => setTimeout(r, 1200));

  // the north end of the same plaza spine the offline takes use, facing south down it
  await page.evaluate(() => window.__ZR_PLAY__.place(0, 8, 0));
  await new Promise((r) => setTimeout(r, 400));
  const started = page.evaluate(
    (s) =>
      window.__ZR_AUDIO__.record(s).then((b) => {
        let bin = '';
        for (let i = 0; i < b.length; i += 0x8000) bin += String.fromCharCode(...b.subarray(i, i + 0x8000));
        return btoa(bin);
      }),
    seconds,
  );
  // Frames are stepped for the whole capture: a headless page asks for no animation frame unless
  // the harness does, and the audio system's parameter update and music scheduler both ride one.
  // Bounded by the WALL clock, because the recorder runs in real time and this page does not
  // animate at 60 Hz, so a fixed frame count does not take a known number of seconds.
  const stepFor = (ms) =>
    page.evaluate(
      async (until, dt) => {
        const t0 = performance.now();
        while (performance.now() - t0 < until) {
          window.__ZR_PLAY__.step(1, dt, false);
          await new Promise((r) => requestAnimationFrame(r));
        }
      },
      ms,
      DT,
    );
  await stepFor(2500);
  await page.keyboard.down('ShiftLeft');
  await page.keyboard.down('KeyW');
  await stepFor((seconds - 3) * 1000);
  await page.keyboard.up('KeyW');
  await page.keyboard.up('ShiftLeft');
  const b64 = await started;
  const stats = await page.evaluate(() => window.__ZR_AUDIO__.stats());
  fs.writeFileSync(path.join(out, 'live.webm'), Buffer.from(b64, 'base64'));
  fs.writeFileSync(path.join(out, 'stats.json'), JSON.stringify(stats, null, 2));
  log(`wrote ${out}/live.webm — ${stats.steps} steps while recording`);
} finally {
  await browser.close();
  await server.close();
}

File: art/audio/2026-09-26-release/live.py
#!/usr/bin/env python3
"""live.py -- the footsteps in a real recording of the real graph, either side of the cut.

    python3 art/audio/2026-09-26-release/live.py --before /tmp/live/before --after /tmp/live/after

`live.mjs` records the master bus while the player runs down the plaza spine. This finds the
footsteps in each recording and reports how hard they land.

It is deliberately a narrow question. Sixteen seconds is far too short a window to resolve a 1.2 Hz
beat, so the pulse is not measured here -- the offline takes do that over sixty-two. What this
answers is the one thing a twin cannot: that the change is in the graph the owner actually hears.

The figure to compare it against is NOT the 4 dB on the trim. A live recording is the master bus,
so every "step peak" in it is a step plus whatever the bed and the music are doing at that instant,
and that sum does not fall by the whole of what the step fell by. Run the same detector over the
offline takes and the size of the effect is plain:

    the offline steps stem   the median step peak falls 4.0 dB -- the trim, exactly
    the offline mix          the same detector sees 2.4 dB

So a live take, which can only ever be the second kind, should land near 2.4 dB. Anything near 0
would mean the cut never reached the shipped path.

Steps are found as peaks in the envelope's rise, at least 150 ms apart (a run's stride is 273 ms
and a heel and its toe are 50-100 ms), the same detector `2026-09-26-perstep` uses. The two
recordings are independent real-time takes of a non-deterministic system, so the two step counts
will not match exactly and the number to read is the median, not any individual step.
"""
import argparse
import math
import os
import wave

import numpy as np


def read(path):
    with wave.open(path, 'rb') as w:
        n, ch, rate = w.getnframes(), w.getnchannels(), w.getframerate()
        raw = w.readframes(n)
    x = np.frombuffer(raw, dtype='<i2').astype(np.float64) / 32768.0
    return (x.reshape(-1, ch).mean(axis=1) if ch > 1 else x), rate


def db(v):
    return 20 * math.log10(max(float(v), 1e-12))


def envelope(x, rate, win=0.004):
    n = max(2, int(win * rate))
    return np.sqrt(np.convolve(x * x, np.ones(n) / n, mode='same'))


def onsets(x, rate, min_gap=0.15):
    env = envelope(x, rate)
    d = np.diff(env, prepend=env[0])
    thr = np.percentile(d, 99.0)
    gap = int(min_gap * rate)
    out, i = [], 0
    while i < len(d):
        if d[i] > thr:
            j = min(len(env), i + int(0.02 * rate))
            out.append(int(i + np.argmax(env[i:j])) if j > i else i)
            i += gap
        else:
            i += 1
    return np.array(out)


def floor_db(x, rate, window=0.05):
    k = int(rate * window)
    x = x[: len(x) // k * k].reshape(-1, k)
    return db(np.percentile(np.sqrt((x**2).mean(axis=1)), 10))


ap = argparse.ArgumentParser()
ap.add_argument('--before', default='/tmp/live/before')
ap.add_argument('--after', default='/tmp/live/after')
# the run starts 3 s in; skip a little more so the first stride is up to speed
ap.add_argument('--from-seconds', type=float, default=4.0)
# what the same detector sees on the offline mix, which is the like-for-like figure
ap.add_argument('--offline-mix-db', type=float, default=-2.4)
a = ap.parse_args()

print('the real graph, recorded while running down the plaza spine\n')
print(f"{'':8} {'steps':>6} {'median step':>13} {'loudest step':>14} {'the floor under them':>21}")
med = {}
for name, folder in (('before', a.before), ('after', a.after)):
    x, rate = read(os.path.join(folder, 'live.wav'))
    x = x[int(a.from_seconds * rate) :]
    at = onsets(x, rate)
    peaks = np.array([np.abs(x[i : i + int(0.04 * rate)]).max() for i in at if i + int(0.04 * rate) < len(x)])
    med[name] = db(np.median(peaks))
    print(f'{name:>8} {len(peaks):>6} {db(np.median(peaks)):>10.1f} dB {db(peaks.max()):>11.1f} dB {floor_db(x, rate):>18.1f} dB')

moved = med['after'] - med['before']
print(f"\na step in the live graph lands {moved:+.1f} dB; the same detector on the offline MIX sees {a.offline_mix_db:+.1f} dB,")
print(f'and on the offline STEPS STEM {-4.0:+.1f} dB, which is the trim exactly. The live take is the first kind.')
